vllm/v1/core/scheduler.py

```diff
@@ -545,6 +601,27 @@ def update_from_output(
             new_logprobs = None
             new_token_ids: List[int] = []
 
+            # Handle guided decoding FSM advancement if applicable
+            # NOTE: For all requests that uses guided decoding, the grammar
+            # should be ready at this point.
+            # PERF: This is currently expensive given that FSM is being
+            # advanced here.
+            if request.use_guided_decoding:
+                grammar = request.grammar
+                assert grammar is not None
+                if len(generated_token_ids) > 1:
+                    logger.error(
+                        "Structured output does not currently support "
+                        "more than one token at a time. Only the first "
+                        "token will be used.")
+                # accept_token advances the FSM
+                accepted = grammar.accept_token(generated_token_ids[0])
+                if not accepted:
+                    logger.error(
+                        "Failed to advance FSM for request %s "
+                        "for tokens %s. Please file an issue.", req_id,
+                        generated_token_ids[0])
```

Comment:
Again, I think we shouldn't have this code directly in the scheduler. This logic should be encapsulated somewhere else.

Reply:
Sure, I moved most of the logic into the guided decoding manager. What's left here is:

```python
            if new_token_ids and request.use_guided_decoding:
                assert request.grammar is not None
                request.grammar.accept_tokens(request.request_id,
                                              new_token_ids)
```